include/deal.II/fe/mapping_isoparametric.h

```diff
@@ -0,0 +1,731 @@
+
+    /**
+     * Second derivative of shape function in quadrature point. See above.
+     */
+    Tensor<2, dim> &
+    second_derivative(const unsigned int qpoint, const unsigned int shape_nr);
+
+    /**
+     * third derivative of shape function in quadrature point. See above.
+     */
+    const Tensor<3, dim> &
+    third_derivative(const unsigned int qpoint,
+                     const unsigned int shape_nr) const;
+
+    /**
+     * third derivative of shape function in quadrature point. See above.
+     */
+    Tensor<3, dim> &
+    third_derivative(const unsigned int qpoint, const unsigned int shape_nr);
+
+    /**
+     * fourth derivative of shape function in quadrature point. See above.
+     */
+    const Tensor<4, dim> &
+    fourth_derivative(const unsigned int qpoint,
+                      const unsigned int shape_nr) const;
+
+    /**
+     * fourth derivative of shape function in quadrature point. See above.
+     */
+    Tensor<4, dim> &
+    fourth_derivative(const unsigned int qpoint, const unsigned int shape_nr);
+
+    /**
+     * Return an estimate (in bytes) for the memory consumption of this object.
+     */
+    virtual std::size_t
+    memory_consumption() const override;
+
+    /**
+     * Values of shape functions. Access by function @p shape.
+     *
+     * Computed once.
+     */
+    std::vector<double> shape_values;
+
+    /**
+     * Values of shape function derivatives. Access by function @p derivative.
+     *
+     * Computed once.
+     */
+    std::vector<Tensor<1, dim>> shape_derivatives;
+
+    /**
+     * Values of shape function second derivatives. Access by function @p
+     * second_derivative.
+     *
+     * Computed once.
+     */
+    std::vector<Tensor<2, dim>> shape_second_derivatives;
+
+    /**
+     * Values of shape function third derivatives. Access by function @p
+     * second_derivative.
+     *
+     * Computed once.
+     */
+    std::vector<Tensor<3, dim>> shape_third_derivatives;
+
+    /**
+     * Values of shape function fourth derivatives. Access by function @p
+     * second_derivative.
+     *
+     * Computed once.
+     */
+    std::vector<Tensor<4, dim>> shape_fourth_derivatives;
+
+    /**
+     * Unit tangential vectors. Used for the computation of boundary forms and
+     * normal vectors.
+     *
+     * This array has `(dim-1) * GeometryInfo::faces_per_cell` entries. The
+     * first GeometryInfo::faces_per_cell contain the vectors in the first
+     * tangential direction for each face; the second set of
+     * GeometryInfo::faces_per_cell entries contain the vectors in the second
+     * tangential direction (only in 3d, since there we have 2 tangential
+     * directions per face), etc.
+     *
+     * Filled once.
+     */
+    std::array<std::vector<Tensor<1, dim>>,
+               GeometryInfo<dim>::faces_per_cell *(dim - 1)>
+      unit_tangentials;
+
+    /**
+     * TODO
+     */
+    const FiniteElement<dim, spacedim> &fe;
```

Comment:
```suggestion
    const SmartPointer<const FiniteElement<dim, spacedim>> fe;
```
It's an annoying change but its the convention we use elsewhere.

Reply:
I would suggest an alternative. I think to create a clone would be safer (after I got a segmentation fault today since the finite element got out scope).

Reply:
I think creating a clone is the right approach.